### ISSUE TITLE
^C no longer interrupts long running shell commands.

### ISSUE THREAD

**Issue by pignatenko** [neovim/neovim#1234]
Ran into this issue with the vcscommand plugin. I didn't set my CVS_RSH value and it tired using RSH to get into the server. vcscommand.vim still uses synchronous calls, but I wasn't able to use ^C to interrupt the shell command, I had to wait for a timeout.

**Comment by khalidchawtany** [comment]
I have to force quit **nvim** from **activity-monitor** every time it hangs and trust me, it hangs a lot, almost making it impossible to work on. PLZ Fix this :(

**Comment by tarruda** [comment]
I just did

```vim
:call system('i=0; while ((i < 5)); do ((i = i + 1)); sleep 1; done')
```

Pressing ctrl+c before the 5 seconds are over is working for me, can you describe steps to reproduce the bug?

**Comment by khalidchawtany** [comment]
@tarruda Sorry for late reply. I cannot say what is causing it. However, it hangs especially when I try to run a command like ```:Ag keyword /**/*.php``` through a **Laravel** application directory. Or, when I  open a file which has very long lines in it, like the **Laravel**  welcome (You've arrived) application page which has an **SVG** image in it. ***The most annoying one is, Sometimes jumping to a next file of the quickfix list makes it hang :(***

I did try the cmd you provided. It works for me too. ***However, I have to press <C-C> more than once (at least twice), else nothing happens***.

**Comment by oni-link** [comment]
In DEBUG mode 
```viml
:call system('sleep 10')
```
can assert when pressing CTRL-C. See https://github.com/neovim/neovim/pull/1365#discussion_r22559122.

**Comment by khalidchawtany** [comment]
I am really sorry for this very* late comment. I have been out of internet service coverage( & Electric too :(  ) because of snow.

If by **Debug** mode you mean the **-D**, I did try the command you provided in debug mode and I was not able to interrupt (using ^C) the command at all (even after the 10 seconds had passed).
```nvim -D -c ":call system('sleep 10')"```

I did try the below command and I was able to interrupt it after a lot of pressing **^C**s then I got the error that I have attached in the screenshot. However, this totally killed the ```nvim``` instance and brought me back to terminal.
```nvim -D -c ":call system('yes')"```


![screen shot 2015-01-13 at 10 33 48](https://cloud.githubusercontent.com/assets/1681606/5717669/c5c3aca6-9b15-11e4-85a9-fd69ee6c2e43.png)


I will update **neovim** and try again to see if it has been fixed in the latest updates.

**Comment by khalidchawtany** [comment]
After a reinstall from **homebrew** sometimes I can interrupt using ^C and there was a time I could't (I had to kill the tmux pane containing it so I don't have a screenshot of it). However, after I interrupt it and ***send :q to quit the running NeoVim instance*** an error is shown.

There was an instance where no error was thrown I have marked in the figure below.
![screen_shot_2015-01-13_at_11_32_42](https://cloud.githubusercontent.com/assets/1681606/5717919/d154f57c-9b18-11e4-84d0-fdd04f6a55ca.png)

**Comment by oni-link** [comment]
Sorry for not being clear. "DEBUG mode" was my way of saying "nvim's build type has to be Debug" (nvim --version|grep Build). Otherwise aborting `:call system('sleep 10')` would show no error for me.

For `:call system('yes')` memory errors are expected (for build type Debug/Release) without the fix in https://github.com/neovim/neovim/pull/1798. Does this fix help with nvim being more responsive when hitting CTRL-C?

**Comment by khalidchawtany** [comment]
I did a reinstal from homebrew, I don't know wether that specific fix is was included or not? However, after this reinstall, I can think that nvim  is more responsive when pressing ^C. Since, now I can interrupt this
```bash
nvim -D -c ":call system('sleep 10')"
```
Which I was not able to do before this reinstall.

**Comment by tarruda** [comment]
Is there are way to install @oni-link's branch from homebrew? This way @khalidchawtany can test if the PR fixes the problem

**Comment by Grimy** [comment]
#1798 was merged, but I still have the same problem on the current master.

    vim  -u NONE +'!yes' +q     # ^C works everytime
    nvim  -u NONE +'!yes' +q    # ^C does nothing

Be prepared to `killall yes` from another terminal if you try the second command.

**Comment by tarruda** [comment]
> #1798 was merged, but I still have the same problem on the current master.

For me the ctrl+c works but it takes one or two seconds, possibly because of the amount of queued reads from libuv

**Comment by Grimy** [comment]
With my '!yes' example, even mashing ^C and waiting a minute does not help in my case. Here’s another test:

    nvim -u NONE +'!while true; echo 42; sleep .1; end' +q

I get the following output after pressing ^C:

	shell returned -1

	Interrupt: Press ENTER or type command to continue42
	42
	42
	42
	42
	42
	42
	42
	42
	42

Which suggests that the command kept running for a full second.

**Comment by tarruda** [comment]
> With my '!yes' example, even mashing ^C and waiting a minute does not help in my case. Here’s another test:

What is your OS/shell?

**Comment by Grimy** [comment]
$ uname -a
    Linux 3.11.10-100.fc18.x86_64 #1 SMP Mon Dec 2 20:28:38 UTC 2013 x86_64 x86_64 x86_64 GNU/Linux

I can reproduce the problem using either fish or bash.

**Comment by tarruda** [comment]
> I can reproduce the problem using either fish or bash.

I can only wonder what is happening, even when I set `sh` to `/bin/bash` (I use zsh) the command is killed after about two seconds(need to fix that though).

Perhaps you'd be willing to debug the problem? You can start by setting a break point [here](https://github.com/neovim/neovim/blob/master/src/nvim/os/job.c#L313) and to see if its being hit.

**Comment by fwalch** [comment]
I get the behavior described by @Grimy if I use the `nvim` I installed system-wide (I'm on Arch Linux). The behavior described by @tarruda occurs if I use an `nvim` compiled manually from Neovim's sources, i.e. run `./build/bin/nvim  -u NONE +'!yes' +q`. So it might be related to either different versions of installed libraries (manual compilation uses built-in dependencies, Arch package uses library packages) or Debug vs Release build (Arch uses `RelWithDebInfo`, btw).

**Comment by tarruda** [comment]
> I get the behavior described by @Grimy if I use the nvim I installed system-wide (I'm on Arch Linux). The behavior described by @tarruda occurs if I use an nvim compiled manually from Neovim's sources, i.e. run ./build/bin/nvim -u NONE +'!yes' +q. So it might be related to either different versions of installed libraries (manual compilation uses dependencies as specified in third-party, Arch package uses library packages) or Debug vs Release build (Arch uses RelWithDebInfo, btw)

@fwalch when you install system-wide, what version of libuv is used?

**Comment by Grimy** [comment]
>Perhaps you'd be willing to debug the problem? You can start by setting a break point here and to see if its being hit.

The breakpoint is indeed hit. After resuming execution with `c`, the job continues unharmed. Single-stepping through job_stop revealed this suspiscious line from libuv:

    if (clock_gettime(clock_id, &t))
        return 0;  /* Not really possible. */

The « Not really possible » code path ends up being executed. Not sure what to make of that.

**Comment by fwalch** [comment]
> @fwalch when you install system-wide, what version of libuv is used?

1.2.0

**Comment by oni-link** [comment]
> For me the ctrl+c works but it takes one or two seconds, possibly because of the amount of queued reads from libuv

I would have expected that a job killed with CTRL-C needs one or two seconds
to terminate because that is the waiting time until a `SIGTERM` or `SIGKILL` are send:
```c
#define TERM_TIMEOUT 1000000000
#define KILL_TIMEOUT (TERM_TIMEOUT * 2)
```

**Comment by tarruda** [comment]
> I would have expected that a job killed with CTRL-C needs one or two seconds
to terminate because that is the waiting time until a SIGTERM or SIGKILL are send:

:+1: I forgot about this detail, but even with this patch the delay still exists(it is shorter though):

```diff
diff --git a/src/nvim/os/job.c b/src/nvim/os/job.c
index 8db5e3c..08642f7 100644
--- a/src/nvim/os/job.c
+++ b/src/nvim/os/job.c
@@ -312,6 +312,8 @@ int job_wait(Job *job, int ms) FUNC_ATTR_NONNULL_ALL
   if (got_int) {
     got_int = false;
     job_stop(job);
+    uv_process_kill(&job->proc, SIGTERM);
+    job->term_sent = true;
     if (ms == -1) {
       // We can only return, if all streams/handles are closed and the job
       // exited.
```

> The « Not really possible » code path ends up being executed. Not sure what to make of that.

I will investigate, but what version of libuv are you using?

**Comment by tarruda** [comment]
> The « Not really possible » code path ends up being executed. Not sure what to make of that.

If `clock_gettime` fails it will return 0, `job_stop_timer_cb` will never send signals to the job, so that is probably the cause.

If you are willing to investigate further, you can check `strerror(errno)` after the os_hrtime call returns, but perhaps the following patch will fix the problem:

```diff
diff --git a/src/nvim/os/job.c b/src/nvim/os/job.c
index 8db5e3c..27ac57f 100644
--- a/src/nvim/os/job.c
+++ b/src/nvim/os/job.c
@@ -272,6 +272,11 @@ void job_stop(Job *job)
   }

   job->stopped_time = os_hrtime();
+
+  if (!job->stopped_time) {
+    job->stopped_time = UINT64_MAX;
+  }
+
   // Close the job's stdin. If the job doesn't close it's own stdout/stderr,
   // they will be closed when the job exits(possibly due to being terminated
   // after a timeout)
@@ -428,7 +433,7 @@ static void job_stop_timer_cb(uv_timer_t *handle)
       ILOG("Sending SIGTERM to job(id: %d)", job->id);
       uv_process_kill(&job->proc, SIGTERM);
       job->term_sent = true;
-    } else if (elapsed >= KILL_TIMEOUT) {
+    } else if (elapsed >= KILL_TIMEOUT || job->stopped_time == UINT64_MAX) {
       ILOG("Sending SIGKILL to job(id: %d)", job->id);
       uv_process_kill(&job->proc, SIGKILL);
     }
```

Can you test it?

**Comment by weitzj** [comment]
I also investigated this issue a bit on OSX 10.10.3 since I have a really long running shell command, and had to kill nvim.

I finally figured out why this happened to me - I have a custom keymapping on <CTRL>-C

`map <C-c> y:e ~/tmpclipboard<CR>P:w !pbcopy<CR><CR>:bdelete!<CR>`

Once I deleted the keymapping, I could reproduce the 1 to 2 seconds timeout as mentioned above:

> For me the ctrl+c works but it takes one or two seconds, possibly because of the amount of queued reads from libuv


For comparison. VIM acknowledges my <CTRL>-C keymapping and aborts the command properly using the message: `Command terminated` 
As it is the case, this message is not used any more in nvim (Only the localizations are still present, e.g. https://github.com/neovim/neovim/blob/master/src/nvim/po/de.po#L7650)

So, I guess neovim should respect the <CTRL>-C keymapping, but override if an external command is running?

See the following 2 gifs for comparing the behaviour in vim and neovim 

![vim_command_terminated](https://cloud.githubusercontent.com/assets/829277/7550466/2a5d09ce-f664-11e4-9e80-f7309b455699.gif)
![neovim_abort_hangs](https://cloud.githubusercontent.com/assets/829277/7550467/2f54c08e-f664-11e4-8950-ac7f46f3ba52.gif)

**Comment by critiqjo** [comment]
>So, I guess neovim should respect the `<CTRL>-C` keymapping, but override if an external command is running?

This not only applies to external commands, after mapping `<C-c>`, you cannot interrupt `:sleep 5` either. So, if what you mean is "override `<C-c>` mapping if Neovim is unresponsive" -- it's a no-go! Because, that would make the behavior of `<C-c>` ambiguous. Neovim may decide to stop a "running" command while the user wished to execute his custom mapping right after the (no-so-long) "run" finishes.

But, if it should only affect "external commands", the user may not know the difference for all it matters (e.g. invokes by a plugin); and besides, external commands (esp. the long running ones) should not be "synchronously" invoked from Neovim anymore, with the arrivals of Terminal Emulator and Jobcontrol API.

In short, remap `<C-c>` only if you know your way around!! I think this is a non-issue as of now.

**Comment by weitzj** [comment]
A short update with some further examples.

Note: I disabled my `<C-c>` mapping.
Note: neovim version: 91fcd3378750080fb94d3030a6fc29a4f86b15e8

0. aborting `:sleep 5` with `<C-c>` works as already mentioned with a 1-2 second delay.

1. Using the ctrlp plugin (https://github.com/kien/ctrlp.vim) it is possible that you are indexing a large directory. Sometimes I want to abort this. In vim I can type `<C-c>` (with or without the mapping) and it will abort the indexing immediately. In neovim I cannot do this. It will not let me abort the indexing.

2. Using the vim-go plugin (https://github.com/fatih/vim-go) it is possible to run you gocode from vim using `go run`. In vim I can stop `go run` using `<C-c>` (with or without the mapping). In neovim this is not possible.

**Comment by aubble** [comment]
I'm kind of confused, is this being fixed in the 0.1 first public release or 0.3?

**Comment by justinmk** [comment]
I guess you're confused by GitHub's history notes, you may as well ignore them. The milestone is clearly listed in the right column.

**Comment by strokirk** [comment]
Maybe related, but in vim I can use `CTRL-C` to abort a long-running command like `:find <tab>` in a large directory, but in Neovim that is not possible. `:find` finishes before neovim accepts any more input.

**Comment by justinmk** [comment]
@strokirk what `:version` ? Works for me on latest (though occasionally segefaults. https://github.com/neovim/neovim/pull/3029 )

**Comment by tarruda** [comment]
> @strokirk what :version ? Works for me on latest (though occasionally segfaults. #3029 

I don't think it is related to #3029, could you open a separate issue and assign to me?

**Comment by Grimy** [comment]
Doesn’t work for me on latest nvim ().

Steps to reproduce
------------------------

* Run `nvim -u NONE +'map <C-C> <Nop>' +'!sleep 5'`
* Hit <kbd>C-c</kbd>

Expected result
--------------------

```^C
Command terminated

Interrupt: Press ENTER or type command to continue```
(Note: this is vim’s behavior. We should either match it, or make a conscious and documented decision to deviate from it on this point)

Actual result
----------------

Nothing happens for five seconds, then Neovim opens.

**Comment by tarruda** [comment]
@Grimy hmm

Should ctrl+c interrupt even if it is remapped?

**Comment by tarruda** [comment]
@Grimy nvm, this happens because vim switches from raw mode and lets the terminal handle ctrl+c.

Does gvim presents the same behavior?

**Comment by Grimy** [comment]
I’ll have to install gvim to check that, I’ll update in a few minutes.

**Comment by tarruda** [comment]
> No, gvim is even worse: in gvim, you can’t interrupt the sleep 5 with C-c at all, even if C-c isn’t remapped. We should definitely avoid matching that, and instead try to match Vim’s behavior, which makes much more sense (to me, at least).

You mean having ctrl+c trigger an interrupt despite being remapped?

**Comment by justinmk** [comment]
I agree, I'm confused about the map step. What is the point?

**Comment by justinmk** [comment]
Oh, it's explained in https://github.com/neovim/neovim/issues/1234#issuecomment-100492190

Ctrl-c should be sent if a job is blocking nvim, even if it is remapped. This does seem reasonable.

**Comment by Grimy** [comment]
Yes. As @weitzj put it:

> neovim should respect the <CTRL>-C keymapping, but override if an external command is running

The benefit of this is that you can remap <kbd>C-c</kbd> to whatever you wish *without* losing the ability to interrupt long-running commands.

As an alternative, it would be okay if there was some other way to interrupt commands. In both Vim and Neovim, mapping <kbd>X</kbd> to <kbd>C-c</kbd> won’t let you interrupt commands with <kbd>X</kbd>.

**Comment by justinmk** [comment]
> override if an external command is running

Just to be clear, this isn't precise enough, because external commands may be running as jobs without blocking. Mapped `ctlr-c` should only send interrupt if the UI is _blocking_.

**Comment by aubble** [comment]
as weitzj mentions in vim-go if you do GoRun you can't stop the program, vim just returns but the program keeps running. makes it annoying if your working with a server, have to kill it each time then run again or use a terminal on the side.

**Comment by tarruda** [comment]
> As an alternative, it would be okay if there was some other way to interrupt commands. In both Vim and Neovim, mapping X to C-c won’t let you interrupt commands with X.

This a consequence of ctrl+c interrupt being handled before mappings are computed. The simplest solution would be to create a `vim_interrupt` API function that is called by the UI when it wants to interrupt, but it has a disadvantage: it would be up to the UI to handle the interrupt mapping, so there would be some code duplication.

**Comment by Grimy** [comment]
> Mapped ctlr-c should interrupt only if the UI is blocking, including non-external processes such as :find.

Yes, this is actually what I meant. Thanks for the clarification.

**Comment by mjacobus** [comment]
+1

Using vdebug plugin  "Waiting for a connection (Ctrl-C to cancel, this message will self destruct in 20 seconds...)

Ctrl+C won't cancel.

**Comment by weitzj** [comment]
Note:

I think this Issue somehow got diverted into talking about how a remapping of `<C-c>` does not quite work, yet, with neovim.

But, even without a remapping, this issue came up for me with the vim-go and CtrlP plugin.
So actually I think this issue should be split up, into two separate issues (which might be related)

- ^C no longer interrupts long running shell commands.
- Remapping ^C overrides default interrupt behaviour

This bug cannot be reproduced with the newest vim-go plugin, since vim-go has now proper neovim support https://github.com/fatih/vim-go/pull/607

I do not know about the CtrlP plugin, but it seems like it might still be "broken" (== neovim behaviour broken)   https://github.com/neovim/neovim/issues/3373#issuecomment-142298114  https://github.com/neovim/neovim/issues/1234#issuecomment-168709967

**Comment by justinmk** [comment]
> Remapping ^C overrides default interrupt behaviour

That should be fixed, assuming you are on 0.1.2 https://github.com/neovim/neovim/pull/3996

**Comment by Grimy** [comment]
Confirmed fixed. Using 75f6817, my repro steps in https://github.com/neovim/neovim/issues/1234#issuecomment-130327842 give the expected result.

**Comment by justinmk** [comment]
As [alluded to](https://github.com/neovim/neovim/issues/1234#issuecomment-64955226) this problem is caused by the fact that nvim never switches to "cooked mode" (https://github.com/neovim/neovim/pull/1613) so CTRL-C does not conveniently set `got_int` for us. 

Instead we must check the input queue for a literal `^C`. This isn't being done effectively in places like `shell.c:append_to_screen_end()`.

At the very least we should check `got_int` in more places so that SIGTERM is respected.
